### PR TITLE
[RFC DRAFT WIP ...] refactor `eval("require")` workaround hack

### PR DESCRIPTION
**Pull request by brodybits**
I noticed that the workaround solution of using eval("require")(...) is repeated throughout `src/common/internal-plugins.js`. I would really favor keeping this kind of a workaround hack in one place for DRY and to decrease the chance of missing places where we may want to improve this solution in the future.

This proposal is triggered by PR #7508, and I hope we can find a way to refactor instances of the workaround hack out of that PR as well.

I am raising this proposal as a DRAFT PR for now, and would be equally happy if this PR is closed in favor of a different solution. I really hope we can find a way to refactor this workaround solution somehow.

/cc @fisker

- ~~I’ve added tests to confirm my change works.~~
- ~~(If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)~~
- ~~(If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.~~
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

Testing:

- [x] I verified that `npx eslint` does not show any errors on the updated modules.
- [x] I verified that `npm test` passes in my work area.

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
This is not how it works, that babel plugin only look for 
```
eval('require')(module)
//              ^^^^^^ this is required
```

**Comment by fisker** [comment]
If you want change that, you need rewrite `scripts/build/babel-plugins/transform-custom-require.js`, or maybe `@rollup/plugin-replace` can replace, not sure.

File: src/common/require-module.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// NOTE: separate const and function to do the following:
+// * execute the `eval` call just once
+// * increase the chance that an easy-to-understand function name
+//   shows up in case of a stack trace
+
+const requireModuleAtPath = eval("require");
+
+function requireModule(path) {
+  return requireModuleAtPath(path);
+}
```
Comment (fisker) [inline src/common/require-module.js]: Or maybe try this

```js
function requireModule(path) {
  return eval("require")(path);
}
```

Comment (fisker) [reply src/common/require-module.js]: For `require.resolve` we can do

```js
requireModule.resolve = eval("require").resolve 
```

Comment (brodybits) [reply src/common/require-module.js]: I tried it, did not seem to get `yarn prepare-release` to pass.

Comment (fisker) [reply src/common/require-module.js]: It  shouldn't

File: src/cli/util.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ const thirdParty = require("../common/third-party");
 const arrayify = require("../utils/arrayify");
 const isTTY = require("../utils/is-tty");
 
+const requireModule = require("../common/require-module");
```
Comment (fisker) [inline src/cli/util.js]: `const requireModule = require;` does this work?

Comment (brodybits) [reply src/cli/util.js]: Seems to work, yes!

Comment (brodybits) [reply src/cli/util.js]: I spoke too soon, prod build did not work.

If we do fix this on the prod build, we could remove our custom Babel `transform-custom-require.js` plugin.

I think the main problem is this error:

```
[error] Dynamic requires are not currently supported by @rollup/plugin-commonjs
```

From some quick research I found an idea in <https://github.com/winstonjs/logform/issues/5#issuecomment-467652075> to use what is now [`@rollup/plugin-node-resolve`](https://www.npmjs.com/package/@rollup/plugin-node-resolve) ([here on GitHub](https://github.com/rollup/plugins/tree/master/packages/node-resolve)). Some very interesting reading in these sections:

- https://github.com/rollup/plugins/tree/master/packages/node-resolve#browser
- https://github.com/rollup/plugins/tree/master/packages/node-resolve#using-with-rollupplugin-commonjs
- https://github.com/rollup/plugins/tree/master/packages/node-resolve#resolving-built-ins-like-fs

I am now starting to wonder if we could do the whole prod build with an updated set of Rollup packages, with minimal or even zero Babel transforms, and minimal custom scripting.

Unfortunately I cannot promise when I will have extra time to look into this further. I would be happy if someone else wants to take this over.

Comment (fisker) [reply src/cli/util.js]: Right, forgot that. `rollup` replace `require` to this error message.

Comment (fisker) [reply src/cli/util.js]: As I said before. `eval('require')` actually a good one.

**Comment by fisker** [comment]
I try to simply that plugin, find I missed something. This plugin not only replace `evale('require')` to `require`, also fix module path https://github.com/prettier/prettier/blob/a858a315e926c21bd1ebfcde70a08419c9472a8c/scripts/build/babel-plugins/transform-custom-require.js#L25


So, this should work.

```js
function requireModule(moduleId) {
	if (moduleId[0] === ".") {
    	eval("require")(moduleId.splice('/').pop())
    }
}
```

But maybe breaks

```js
eval('require')(variable)
```

File: src/cli/util.js (hunk 2 above)
Comment (brodybits) [reply src/cli/util.js]: > `eval('require')` actually a good one.

Done but not working. I get this kind of an error on prod test:

```
Cannot find module '../language-html/parser-html' from 'index.js'
```

I tried hacking `scripts/build/babel-plugins/transform-custom-require.js` but with no luck.

The custom Babel plugin assumes that we are using `eval('require')` in every case where we need the transform, which is violated by this proposal.

I will now try another idea.

Comment (fisker) [reply src/cli/util.js]: I'm `eval('require')` in the codebase is actually a good solution.

**Comment by fisker** [comment]
my code in https://github.com/fisker/prettier/pull/302 is the way how to replace `eval("require")` to `require`, but if we do `requireModule('./foo')`, we can't locate the argument string now, so no way to replace that string

**Comment by fisker** [comment]
I think we shouldn't try anything more... All solutions I can think are less good than current solution .

**Comment by brodybits** [comment]
OK I am putting this on hold for now.

File: scripts/build/babel-plugins/transform-custom-require.js [hunk 3]
```diff
@@ -19,37 +19,22 @@ module.exports = function(babel) {
...
     }
   };
 
   function isEvalRequire(node) {
     return (
       t.isCallExpression(node) &&
-      t.isIdentifier(node.callee, { name: "eval" }) &&
-      node.arguments.length === 1 &&
-      t.isLiteral(node.arguments[0], { value: "require" })
+      t.isIdentifier(node.callee, { name: "requireModule" })
```
Comment (fisker) [inline scripts/build/babel-plugins/transform-custom-require.js]: Don't try this way, even it works, we need hard code this function name.

**Comment by thorn0** [comment]
I unintentionally closed this and several other PRs because I deleted the `next` branch, but looks like a good solution wasn't found here and it's not really clear whether there is a problem to solve in the first place, so let's keep this PR closed.

**Pull request closed without merging**